sw/device/lib/crypto/impl/ecc/ecdsa_p256.c

```diff
@@ -33,111 +34,113 @@ static const otbn_addr_t kOtbnVarEcdsaD0 = OTBN_ADDR_T_INIT(p256_ecdsa, d0);
 static const otbn_addr_t kOtbnVarEcdsaD1 = OTBN_ADDR_T_INIT(p256_ecdsa, d1);
 static const otbn_addr_t kOtbnVarEcdsaXr = OTBN_ADDR_T_INIT(p256_ecdsa, x_r);
 
-/* Mode is represented by a single word, 1 for sign and 2 for verify */
-static const uint32_t kOtbnEcdsaModeNumWords = 1;
-static const uint32_t kOtbnEcdsaModeSign = 1;
-static const uint32_t kOtbnEcdsaModeVerify = 2;
+/* Mode is represented by a single word. See `p256_ecdsa.s` for values. */
+static const uint32_t kOtbnEcdsaModeWords = 1;
+static const uint32_t kOtbnEcdsaModeKeygen = 0x3d4;
+static const uint32_t kOtbnEcdsaModeSign = 0x15b;
+static const uint32_t kOtbnEcdsaModeVerify = 0x727;
 
-// TODO: This implementation waits while OTBN is processing; it should be
-// modified to be non-blocking.
-status_t ecdsa_p256_sign(const ecdsa_p256_message_digest_t *digest,
-                         const ecdsa_p256_private_key_t *private_key,
-                         ecdsa_p256_signature_t *result) {
+status_t ecdsa_p256_keygen_start(void) {
   // Load the ECDSA/P-256 app. Fails if OTBN is non-idle.
   HARDENED_TRY(otbn_load_app(kOtbnAppEcdsa));
 
-  // Set mode so start() will jump into p256_ecdsa_sign.
-  HARDENED_TRY(otbn_dmem_write(kOtbnEcdsaModeNumWords, &kOtbnEcdsaModeSign,
+  // Set mode so start() will jump into keygen.
+  HARDENED_TRY(otbn_dmem_write(kOtbnEcdsaModeWords, &kOtbnEcdsaModeKeygen,
+                               kOtbnVarEcdsaMode));
+
+  // Start the OTBN routine.
+  return otbn_execute();
+}
+
+status_t ecdsa_p256_keygen_finalize(p256_masked_scalar_t *private_key,
+                                    p256_point_t *public_key) {
+  // Spin here waiting for OTBN to complete.
+  HARDENED_TRY(otbn_busy_wait_for_done());
+
+  // Read the masked private key from OTBN dmem.
+  HARDENED_TRY(otbn_dmem_read(kP256MaskedScalarShareWords, kOtbnVarEcdsaD0,
```

Comment:
nit: Is there going to be a case where they private key is not extracted from otbn? e.g. keygen or derive key and leave it in place for `sign_start`?

Reply:
I was thinking about the attestation flow with sideloaded keys, and the main issue I see is that: 

1. We generate UDS from the CreatorRootKey.
2. We have to advance the keymgr state to OwnerIntermediateKey (CDI_attestation) to generate the CDI public key.
3. We sign CDI with UDS. 

So there is an issue in 3) because the state of the keymgr won't allow us to regenerate UDS. 

In general it would be nice to extend the idea of returning key handles from OTBN whenever possible, to improve isolation with software, but that is a discussion for a future release. 

Thanks for the context!

Reply:
Ah, that makes sense. I think for attestation we can make a different interface, since my impression was that all the ROM/ROM_EXT crypto was going to be essentially separate from the OS-visible cryptolib (i.e. shared OTBN program but separate C interfaces and binaries, like for RSA). Or is that not the case for ROM_EXT?

Reply:
Hi Jade that is correct for ROM_EXT, but in the future, the OS will also want to have the ability to generate attestation signatures. We can split the problem into two and handle the ROM_EXT case first. Later on we can define the requirements for the OS and come up with a solution.

Reply:
Sounds good. I'd also be happy with making attestation its own function in the API.